### ISSUE TITLE
Site that has been 'valet unsecure'd still serves HTTP URLs with 301 to HTTPS

### ISSUE THREAD

**Issue by benlk** [laravel/valet#878]
I'm using valet with a WordPress install, where I'm trying to test some things related to the [WordPress AMP plugin](https://wordpress.org/plugins/amp/). The AMP plugin contains functionality that uses PHP's curl to fetch the site's own pages, in order to validate the page's markup for AMP compliance.

When the site had been `valet secure`d, the AMP validation plugin complained that it was unable to connect to the server, because of a certificate validation failure.

I ran `valet unsecure` to try to get valet to stop serving the site over HTTPS, and `wp search-replace "https://trust.test" "http://trust.test"` to make sure that the WordPress site was representing itself in links as an HTTP site rather than an HTTPS site. Even after `valet restart` and clearing browser caches, valet continues to serve HTTP URLs with a 301 redirect to the HTTPS location.

My questions are as follows:

- Is there a way to insert a certificate from `valet secure` into PHP curl's cert store so that PHP's curl functionality trusts the self-signed certificates issued by `valet secure`?
- Is there a way to get valet to stop serving 301s?

**Comment by drbyte** [comment]
Valet is only serving the redirect if you have a .conf file for that site that contains the redirect. When running valet secure it creates exactly that sort of conf file for the site, but removes it when running valet unsecure. If for some reason it didn't remove it, you can remove it manually in `~/.config/valet/Nginx`

To make curl aware of Valet's certificate, you can create a simple file in your home directory:
`~/.curlrc` containing one line (substitute your username):
`cacert=/Users/YOUR_USERNAME_HERE/.config/valet/CA/LaravelValetCASelfSigned.pem`

**Comment by benlk** [comment]
Setting the cacert in `~/.curlrc` doesn't seem to fix the problem where PHP's curl functionality doesn't recognize the validity of the certificate.

I also added `curl.cainfo = /Users/MY_USERNAME_HERE/.config/valet/CA/LaravelValetCASelfSigned.pem` to `/usr/local/etc/php/7.3/php.ini`, and restarted valet, but I'm not getting success. 

The plugin tells me:

> Failed to fetch URL(s) to validate. This may be due to a request timeout. Please check your server's PHP error logs; to do this you may need to enable WP_DEBUG_LOG.

But there's nothing in any WordPress logs concerning failure to fetch, so I'm also not sure whether this may be a bug in the AMP plugin. ☹️

**Comment by drbyte** [comment]
It could indeed be something about the AMP plugin.

I've also seen a few reported situations of Wordpress issues where it relies on something with curl to do background processes, where whatever it's doing for that communication just "doesn't".

If it's a curl thing, then fixing it requires identifying exactly "which" curl process is really doing the communication, and identifying where "that" curl process's configs are found, and correcting those, particularly in telling them where to find the validating Certificate Authority used by Valet.

**Comment by benlk** [comment]
Further research in https://github.com/ampproject/amp-wp/issues/3795 pointed me to the Site Health page at `/wp-admin/site-health.php`, which even after deactivating all plugins reports:

> The REST API encountered an error:
> 
> The REST API is one way WordPress, and other applications, communicate with the server. One example is the block editor screen, which relies on this to display, and save, your posts and pages.
> 
> The REST API request failed due to an error.
> Error: [] cURL error 28: Connection timed out after 10070 milliseconds
> 
> ---
> 
> Your site could not complete a loopback request.
> 
> Loopback requests are used to run scheduled events, and are also used by the built-in editors for themes and plugins to verify code stability.
> 
> The loopback request to your site failed, this means features relying on them are not currently working as expected.
> Error: [] cURL error 28: Connection timed out after 10073 milliseconds

**Comment by benlk** [comment]
The loopback request failure is resolved by making sure that `127.0.0.1` is included in my computer's DNS config, so that when system libraries try to resolve this site they get the right IP address for the domain name.

**Comment by benlk** [comment]
This error no longer exists:

> Failed to fetch URL(s) to validate. This may be due to a request timeout. Please check your server's PHP error logs; to do this you may need to enable WP_DEBUG_LOG.

My site is able to visit itself now. My root problem is resolved: the plugin works now, even over HTTPS.

---

However.

The reported issue, where after `valet unsecure` the site is still 301ing to HTTPS URLs, continues to exist. There's no site-specific config item in `~/.config/valet/Nginx/`, and the HTTPS certificate used is for another site in `valet parked`, one which is marked as having an SSL certificate.

**Comment by drbyte** [comment]
I'm glad you sorted the curl thing out!

> the site is still 301ing to HTTPS URLs, continues to exist.

I'm wondering if it's your browser doing the redirect, perhaps because it's remembered that the site was served over https previously.  While Valet doesn't implement HSTS (auto-redirect to https because of an apache/nginx "policy"), I've sometimes seen browsers do that on their own.
To rule out the browser, you could probably use Postman or something to inspect the actual http negotiation process.
(I'm sure you've already edited any configs in your app to tell them to use http only, but I'll mention it just in case.)

But ... if the curl issue is resolved, do you still need to "not" use https?

**Comment by benlk** [comment]
The curl issue is still resolved, and so I don't need to use HTTP instead of HTTPS.

The 301 behavior was apparent even in different browsers and after dumping caches.

Other non-HTTPS valet sites that have never been `valet secure`d are also currently redirecting to their HTTPS versions, and being served with an HTTPS cert from a `valet secure` site. (Not the first site alphabetically or the most-recently-secured site, but consistently a specific site.)

**Comment by drbyte** [comment]
> Other non-HTTPS valet sites that have never been `valet secure`d are also currently redirecting to their HTTPS versions, and being served with an HTTPS cert from a `valet secure` site. (Not the first site alphabetically or the most-recently-secured site, but consistently a specific site.)

Yes, the serving of the "first available certificate" to an https request is normal. Particularly when things like Valet are designed to do some degree of wildcarding for simplicity of dev environments.

I'm curious about why it's doing the https redirection on you. If it's coming from Valet itself I'd be a bit surprised (unless you've changed the default nginx configs to do the redirect, but it doesn't sound like you tinkered with that).  But if Postman (or other header-inspecting-capable tool) reveals the redirect being initiated from Nginx then it'd be worth digging in there.

**Comment by benlk** [comment]
It looks like all sites that have not been explicitly issued certificates with `valet secure`, when visited via HTTP, are being served with a 301 to the equivalent HTTPS URL.

The cert used for these sites is, after poking around in `~/.config/valet/Certificates` with `ls -alS`, the smallest `.key` file.

Here's a site which `valet parked` and `ls ~/.config/valet/Certificates` shows has not been issued a certificate, being redirected to HTTPS:

```
$ curl -v http://phpinfo.test
* Rebuilt URL to: http://phpinfo.test/
*   Trying 127.0.0.1...
* TCP_NODELAY set
* Connected to phpinfo.test (127.0.0.1) port 80 (#0)
> GET / HTTP/1.1
> Host: phpinfo.test
> User-Agent: curl/7.54.0
> Accept: */*
> 
< HTTP/1.1 301 Moved Permanently
< Server: nginx/1.15.6
< Date: Fri, 20 Dec 2019 19:39:47 GMT
< Content-Type: text/html
< Transfer-Encoding: chunked
< Connection: keep-alive
< Location: https://phpinfo.test/
< 
<html>
<head><title>301 Moved Permanently</title></head>
<body>
<center><h1>301 Moved Permanently</h1></center>
<hr><center>nginx/1.15.6</center>
</body>
</html>
* Connection #0 to host phpinfo.test left intact
```

**Comment by drbyte** [comment]
Let's compare your `/usr/local/etc/nginx/valet/valet.conf` with the Valet default stored in `~/.composer/vendor/laravel/valet/cli/stubs/valet.conf`

Valet doesn't do a redirect unless the site is secured, which it does by copying the `secure.valet.conf` from that stubs dir, and substituting the site details when generating Certificates.

**Comment by benlk** [comment]
```diff
$ diff ~/.composer/vendor/laravel/valet/cli/stubs/valet.conf /usr/local/etc/nginx/valet/valet.conf 
2c2
<     listen 127.0.0.1:80 default_server;
---
>     listen 80 default_server;
7c7
<     location /VALET_STATIC_PREFIX/ {
---
>     location /41c270e4-5535-4daa-b23e-c269744c2f45/ {
14c14
<         rewrite ^ "VALET_SERVER_PATH" last;
---
>         rewrite ^ "/Users/user/.composer/vendor/laravel/valet/server.php" last;
17,19d16
<     location = /favicon.ico { access_log off; log_not_found off; }
<     location = /robots.txt  { access_log off; log_not_found off; }
< 
21c18
<     error_log "VALET_HOME_PATH/Log/nginx-error.log";
---
>     error_log "/Users/user/.config/valet/Log/nginx-error.log";
23c20
<     error_page 404 "VALET_SERVER_PATH";
---
>     error_page 404 "/Users/user/.composer/vendor/laravel/valet/server.php";
25c22
<     location ~ [^/]\.php(/|$) {
---
>     location ~ \.php$ {
27,28c24,25
<         fastcgi_pass "unix:VALET_HOME_PATH/valet.sock";
<         fastcgi_index "VALET_SERVER_PATH";
---
>         fastcgi_pass "unix:/Users/user.config/valet/valet.sock";
>         fastcgi_index "/Users/user/.composer/vendor/laravel/valet/server.php";
30,31c27
<         fastcgi_param SCRIPT_FILENAME "VALET_SERVER_PATH";
<         fastcgi_param PATH_INFO $fastcgi_path_info;
---
>         fastcgi_param SCRIPT_FILENAME "/Users/user/.composer/vendor/laravel/valet/server.php";
```

**Comment by drbyte** [comment]
How much have you customized anything related to Valet? (php, nginx, dnsmasq)

I ask because I'd like to suggest upgrading to the latest Valet version and running `valet install` to reconfigure things to defaults.

**Comment by drbyte** [comment]
@benlk Is this still an issue? Or can this be closed?

**Comment by benlk** [comment]
I don't believe I had customized anything related to Valet in terms of php, nginx, or dnsmasq.

I'm not currently working on the project where this was an issue, and as mentioned in https://github.com/laravel/valet/issues/878#issuecomment-568059676 I was able to get the project to work around Valet's behavior.

I don't know if this is still an issue, however, and don't have time to `valet install`, sorry.
